lib/ansible/modules/windows/win_zip.ps1

```diff
@@ -0,0 +1,253 @@
+                }
+            }
+        }
+    }
+    else {
+        $item = Get-Item -LiteralPath $src
+        if ($item.Length -ge $max_size) {
+            ($large_files.Value).Add($item.FullName) | Out-Null
+        }
+    }
+}
+
+function New-ZipFile {
+	param(
+	[Parameter(Mandatory=$True)]
+	[string]$src,
+	[Parameter(Mandatory=$True)]
+	[string]$dest,
+	[Parameter(Mandatory=$True)]
+	[CompressionUtils]$utils_available
+	)
+
+    switch ($utils_available)
+    {
+        "PowerShell" {
+            try {
+                if (-not $check_mode) {
+                    Compress-Archive -LiteralPath $src -DestinationPath $dest -CompressionLevel Optimal -ErrorAction Stop
+                }
+                #$result.changed = $true
+            }
+            catch {
+                Remove-BrokenZip $dest
+                Fail-Json -obj $result -message "Error creating zip file '$dest': $($_.Exception.Message)"
+            }
+        }
+        "DotNet" {
+            try {
+                # add type here is only needed for force testing dotnet in the calling main code
+                Add-Type -AssemblyName System.IO.Compression.FileSystem -ErrorAction Stop| Out-Null
+                Add-Type -AssemblyName System.IO.Compression -ErrorAction Stop | Out-Null
+
+                if (Test-Path -LiteralPath $src -PathType Container)
+                {
+                    if (-not $check_mode) {
+                        [System.IO.Compression.ZipFile]::CreateFromDirectory($src, $dest)
+                    }
+                    #$result.changed = $true
+                }
+                else {
+                    if (-not $check_mode) {
+                        $archive = [System.IO.Compression.ZipFile]::Open($dest, [System.IO.Compression.ZipArchiveMode]::Create, [System.Text.Encoding]::UTF8)
+                        $filename = (Get-Item -LiteralPath $src).Name
+                        [System.IO.Compression.ZipFileExtensions]::CreateEntryFromFile($archive, $src, $filename, 0) | Out-Null
+                    }
+                    #$result.changed = $true
+                }
+            }
+            catch {
+                # If zip is opened but something fails afterwards, it will be left on disk. Need to cleanup
+                # Duplicated here and not just finally because we have to dispose it before we can delete it
+                # or the unmanaged file is locked
+                if ((Get-Variable -Name archive -ErrorAction SilentlyContinue) -and ($null -ne $archive)) {
+                    $archive.Dispose()
+                }
+                Remove-BrokenZip $dest
+                Fail-Json -obj $result -message "Error creating zip file '$dest': $($_.Exception.Message)"
+
+            }
+            finally {
+                # Call dispose to get rid of the unmanaged object
+                if ((Get-Variable -Name archive -ErrorAction SilentlyContinue) -and ($null -ne $archive)) {
```

Comment:
You can simplify this by defining archive as null before the try section and just changing this to

```
if ($archive) {
    $archive.Dispose()
}
```